src/v/cloud_storage/remote_partition.cc

```diff
@@ -523,27 +523,40 @@ class partition_record_batch_reader_impl final
                 co_return;
             }
 
+            // Out of range queries are unexpected. The caller must take care
+            // to send only valid queries to remote_partition. I.e. the fetch
+            // handler does such validation. Similar validation is done inside
+            // remote partition.
+            //
+            // Out of range at this point is due to a race condition or due to
+            // a bug. In both cases the only valid action is to throw an
+            // exception and let the caller deal with it. If the caller doesn't
+            // handle it it leads to a closed kafka connection which the
+            // end clients retry.
             if (
               cur.error() == error_outcome::out_of_range
               && ss::visit(
                 query,
-                [&](model::offset) { return false; },
+                [&](model::offset) {
+                    vassert(
+                      false,
+                      "Unreachable code. Remote partition doesn't know how to "
+                      "handle model::offset queries.");
+                    return false;
```

Comment:
Where should I look to verify that this code is actually unreachable? May be worth adding a comment in case it ever fires

Reply:
Tests? :)